### PR TITLE
feat(dataframe): design the write related interface of DataFrame

### PR DESCRIPTION
**Pull request by AChenQ**

### CODE REVIEW

**Comment by GravitiRob** [comment]
Task linked: [CU-2dn08t1 [parent] SDK 接口设计 —— 写入数据](https://app.clickup.com/t/2dn08t1)

File: graviti/dataframe/frame.py [hunk 1]
```diff
@@ -476,3 +470,28 @@ def sample(self, n: Optional[int] = None, axis: Optional[int] = None) -> "DataFr
 
     def info(self) -> None:
         """Print a concise summary of a DataFrame."""
+
+    def extend(self, objs: "DataFrame") -> None:
```
Comment (linjiX) [inline graviti/dataframe/frame.py:474]: should `extend` supports the pylist input?

File: graviti/dataframe/indexing.py [hunk 2]
```diff
@@ -36,6 +36,21 @@ def __getitem__(self, key: Iterable[int]) -> "SeriesBase[_T]":
     def __getitem__(self, key: Union[int, Iterable[int]]) -> Any:
         return self.obj._getitem_by_location(key)
 
+    @overload
+    def __setitem__(self, key: int, value: Iterable[Any]) -> None:
+        pass
```
Comment (linjiX) [inline graviti/dataframe/indexing.py:41]: use `...` for overload method

File: graviti/dataframe/indexing.py [hunk 3]
```diff
@@ -36,6 +36,21 @@ def __getitem__(self, key: Iterable[int]) -> "SeriesBase[_T]":
...
+
+    @overload
+    def __setitem__(self, key: Union[Iterable[int], slice], value: Iterable[Iterable[Any]]) -> None:
```
Comment (linjiX) [inline graviti/dataframe/indexing.py:44]: we do not support the `Iterable[int]` input this release.

File: graviti/dataframe/frame.py (hunk 1 above)
Comment (linjiX) [reply graviti/dataframe/frame.py:474]: that means DataFrame.extend add columns, and DataFrame.loc.extend add rows?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py (hunk 1 above)
Comment (linjiX) [reply graviti/dataframe/frame.py:474]: DataFrame columns more like a dict, and dict extend should be `update`, so I think `DataFrame.extend` is not a good name.

File: graviti/dataframe/indexing.py [hunk 4]
```diff
@@ -80,6 +110,55 @@ def __getitem__(self, key: Iterable[int]) -> "DataFrame":
...
+        pass
+
+    def __setitem__(
+        self,
+        key: Union[int, Iterable[int], slice],
+        value: Union[Iterable[Any], Iterable[Iterable[Any]]],
+    ) -> None:
+        pass
+
+    def extend(self, objs: Iterable[Union[Dict["str", Any], "DataFrame"]]) -> None:
```
Comment (graczhual) [inline graviti/dataframe/indexing.py:128]: why we need to add extend rows to loc and iloc, and they look exactly the same. Could we add extend_rows to dataframe?

File: graviti/dataframe/indexing.py [hunk 5]
```diff
@@ -36,6 +36,21 @@ def __getitem__(self, key: Iterable[int]) -> "SeriesBase[_T]":
     def __getitem__(self, key: Union[int, Iterable[int]]) -> Any:
         return self.obj._getitem_by_location(key)
 
+    @overload
+    def __setitem__(self, key: int, value: Iterable[Any]) -> None:
```
Comment (Lee-000) [inline graviti/dataframe/indexing.py:40]: Why the value must be `Iterable[Any]`?

File: graviti/dataframe/indexing.py [hunk 6]
```diff
@@ -101,3 +180,52 @@ def __getitem__(self, key: Iterable[int]) -> "DataFrame":
...
+        pass
+
+    def __setitem__(
+        self,
+        key: Union[_T, slice],
+        value: Union[Iterable[Any], Iterable[Iterable[Any]]],
+    ) -> None:
+        pass
+
+    def extend(self, objs: Iterable[Union[Dict["str", Any], "DataFrame"]]) -> None:
```
Comment (Lee-000) [inline graviti/dataframe/indexing.py:199]: Why the double quotes around `str`?

File: graviti/dataframe/column/series.py [hunk 7]
```diff
@@ -83,6 +83,9 @@ def __getitem__(self, key: Union[int, Iterable[int]]) -> Any:
 
         return self._data[integer_location]
 
+    def __setitem__(self, key: str, value: Any) -> None:
```
Comment (Lee-000) [inline graviti/dataframe/column/series.py:86]: Why the type of `value` here is different from those in indexing?

File: graviti/dataframe/indexing.py [hunk 8]
```diff
@@ -80,6 +110,21 @@ def __getitem__(self, key: Iterable[int]) -> "DataFrame":
...
+        ...
+
+    @overload
+    def __setitem__(self, key: slice, value: Iterable[Iterable[Any]]) -> None:
+        ...
+
+    def __setitem__(
+        self,
+        key: Union[int, slice],
+        value: Union[Iterable[Any], Iterable[Iterable[Any]]],
```
Comment (linjiX) [inline graviti/dataframe/indexing.py:124]: what is the meaning of the type `Iterable[Iterable[Any]]`?

File: graviti/dataframe/frame.py [hunk 9]
```diff
@@ -476,3 +470,37 @@ def sample(self, n: Optional[int] = None, axis: Optional[int] = None) -> "DataFr
...
+
+        Raises:  # noqa: DAR402
+            ValueError: When the key or column name of objs not exists in self dataframe.
+
+        Examples:
+        >>> df = DataFrame([
+                {"filename": "a.jpg", "box2ds": {"x": 1, "y": 1}},
+                {"filename": "b.jpg", "box2ds": {"x": 2, "y": 2}},
+            ])
+        >>> df.loc.extend([{"filename": "c.jpg", "box2ds": {"x": 3, "y": 3}])
```
Comment (graczhual) [inline graviti/dataframe/frame.py:488]: df.extend?

File: graviti/dataframe/indexing.py [hunk 10]
```diff
@@ -101,3 +176,18 @@ def __getitem__(self, key: Iterable[int]) -> "DataFrame":
...
+        ...
+
+    @overload
+    def __setitem__(self, key: slice, value: Iterable[NestedDict[str, Any]]) -> None:
+        ...
+
+    def __setitem__(
+        self,
+        key: Union[_T, slice],
+        value: Union[NestedDict[str, Any], Iterable[NestedDict[str, Any]]],
```
Comment (linjiX) [inline graviti/dataframe/indexing.py:191]: can the value be a normal scalar?

File: graviti/dataframe/frame.py [hunk 11]
```diff
@@ -476,3 +470,37 @@ def sample(self, n: Optional[int] = None, axis: Optional[int] = None) -> "DataFr
 
     def info(self) -> None:
         """Print a concise summary of a DataFrame."""
+
+    def extend(self, objs: Union[Iterable[Dict[str, Any]], Iterable["DataFrame"]]) -> None:
```
Comment (linjiX) [inline graviti/dataframe/frame.py:474]: why `Iterable["DataFrame"]`?
should it be just "DataFrame"?

File: graviti/dataframe/indexing.py [hunk 12]
```diff
@@ -7,56 +7,116 @@
...
-    from graviti.dataframe.series import SeriesBase
+
+    _S = TypeVar("_S", RowSeries, ColumnSeries)
 
 _T = TypeVar("_T", int, str)
 
 
-class SeriesILocIndexer(Generic[_T]):
-    """Index class for Series.iloc."""
+class SeriesILocIndexerBase(Generic["_S"]):
```
Comment (linjiX) [inline graviti/dataframe/indexing.py:22]: since the SeriesBase is been removed, do we still need IndexerBase?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/indexing.py (hunk 12 above)
Comment (AChenQ) [reply graviti/dataframe/indexing.py:22]: `RowSeriesILocIndexer` and `ColumnSeriesILocIndexer` have the same function `__getitem__`

Comment (linjiX) [reply graviti/dataframe/indexing.py:22]: Are they really the same?
Can you provide some example code?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/indexing.py (hunk 12 above)
Comment (AChenQ) [reply graviti/dataframe/indexing.py]: Remove the `IndexerBase` now.

File: graviti/dataframe/indexing.py (hunk 10 above)
Comment (AChenQ) [reply graviti/dataframe/indexing.py:213]: Not support now.

File: graviti/dataframe/frame.py [hunk 13]
```diff
@@ -35,28 +35,22 @@ class DataFrame:
...
-
-        >>> data = {"col1": [1, 2], "col2": [3, 4]}
-        >>> df = DataFrame(data=data)
-        >>> df
-            col1 col2
-        0   1    3
-        1   2    4
+        Constructing DataFrame from list.
 
         >>> df = DataFrame(
```
Comment (graczhual) [inline graviti/dataframe/frame.py:47]: duplicated

File: graviti/dataframe/frame.py [hunk 14]
```diff
@@ -35,28 +35,22 @@ class DataFrame:
...
-                {
-                    "filename": ["a.jpg", "b.jpg", "c.jpg"],
-                    "box2ds": {"x": [1, 2, 3], "y": [1, 2, 3]},
-                }
+        >>> df = DataFrame(
+                [
+                    {"filename": "a.jpg", "box2ds": {"x": 1, "y": 1}},
+                    {"filename": "b.jpg", "box2ds": {"x": 2, "y": 2}},
+                    {"filename": "c.jpg", "box2ds": {"x": 3, "y": 3}},
+                ]
```
Comment (graczhual) [inline graviti/dataframe/frame.py:53]: use `...`?

**Review by graczhual — CHANGES_REQUESTED** [review]

**Review by graczhual — APPROVED** [review]

File: graviti/dataframe/indexing.py [hunk 15]
```diff
@@ -5,59 +5,141 @@
...
     from graviti.dataframe.dataframe import DataFrame
     from graviti.dataframe.row.series import Series as RowSeries
-    from graviti.dataframe.series import SeriesBase
 
 _T = TypeVar("_T", int, str)
 
 
-class SeriesILocIndexer(Generic[_T]):
-    """Index class for Series.iloc."""
+class RowSeriesILocIndexer:
```
Comment (linjiX) [inline graviti/dataframe/indexing.py:20]: since the index base class is removed, should we move the index classes into row and column folder?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/indexing.py [hunk 16]
```diff
@@ -5,62 +5,56 @@
...
     from graviti.dataframe.dataframe import DataFrame
     from graviti.dataframe.row.series import Series as RowSeries
-    from graviti.dataframe.series import SeriesBase
 
 _T = TypeVar("_T", int, str)
 
 
-class SeriesILocIndexer(Generic[_T]):
-    """Index class for Series.iloc."""
+class DataFrameILocIndexer:
```
Comment (linjiX) [inline graviti/dataframe/indexing.py:19]: why the file name is `indexing` not `indexer`?

**Review by linjiX — APPROVED** [review]

File: graviti/dataframe/frame.py [hunk 17]
```diff
@@ -283,11 +276,11 @@ def shape(self) -> Tuple[int, int]:
...
-                    {
-                        "filename": ["a.jpg", "b.jpg", "c.jpg"],
-                        "box2ds": {"x": [1, 2, 3], "y": [1, 2, 3]},
-                    }
-                )
+            ...     {
+            ...         "filename": ["a.jpg", "b.jpg", "c.jpg"],
+            ...         "box2ds": {"x": [1, 2, 3], "y": [1, 2, 3]},
+            ...     }
+            ... )
```
Comment (Lee-000) [inline graviti/dataframe/frame.py:283]: The construction in the example is different from the one in the `init` example? Should it be updated?

**Review by Lee-000 — APPROVED** [review]

**Pull request merged**